### DISCUSSION TITLE
Copy & Advance skips differences when moved block detection is on

### DISCUSSION THREAD

**Discussion by Khoji** [WinMerge/winmerge#1234] [Q&A] [answered]
Hopefully this hasn't been asked yet, I couldn't find it on a search: 

I use WinMerge a lot for updating script files in the original source after editing the test version. There are usually a lot of changes to update, which I work through with the Copy & Advance tool. However, this frequently skips over some changes if I just click through with the Copy & Advance button. Sometimes this is just whitespace, although I have it configured to compare that, but not always. It often skips over real changes, and if I'm also not changing everything it's easy to miss something important that should have been copied.

I've done a lot of testing with different settings and it seems to be related to moved block detection. As far as I can see, it doesn't happen when MBD is off, but it does when it is on, even if there are no blocks that have really been moved. It may also be related to the presence of deleted lines, but not always. 

Is this a known issue? I can work around it by switching MBD off, but it's a nice feature to have.

**Comment by sdottaka** [comment, marked as answer]
This bug has been fixed with the following commit. Thank you for the report. 
https://github.com/WinMerge/winmerge/commit/274a841eedfb54508f8420d36ae16b03b62d1d04

**Reply by Khoji** [reply]
Cool, thanks for the lightning fast response!

**Reply by pridmorej** [reply]
This is still a bug in 2.16.30, and Copy Right and Advance skips blocks regardless of the "Enable moved block detection" setting.

**Comment by sdottaka** [comment]
> This is still a bug in 2.16.30, and Copy Right and Advance skips blocks regardless of the "Enable moved block detection" setting.

I could not reproduce it with 2.16.30 or the latest 2.16.34.
Please tell me the conditions to reproduce it.

**Comment by pridmorej** [comment]
Hiya Takashi, thanks for replying.

I'm using winmerge as the merge tool for git, so it is opened with several
parameters for local, base and remote files and also the -o parameter for
the output file, plus params for captions and skipping prompts etc.  Left
and right panes are read-only (set via parameters).  Presumably the base
file is the one with the merge conflict markers and local and remote do not
contain any merge conflict markers.  When I click save, the middle pane's
header changes to the name of the output file as expected.  That all seems
to work pretty well.

Most of the changes I'm trying to keep are on the left (local) so I'm using
the 'copy left and advance' button, and when there are several differences
then often (but not always) certain differences are skipped as I copy left
and advance down.  The skipped differences can be single line or multiline
(makes no difference).

Different issue, but it would also be a great improvement if the 'copy
left/right and advance' buttons were enabled when the middle pane has
focus.  I find it a pain to have to mouse click between the left and right
panes when I want to copy a change from one side and then a different
change from the other side.

Happy to share any other details to help.

Kind regards
Jeremy

**Comment by sdottaka** [comment]
Thank you for the information.
I don't know the conditions, but when I had three files open, an unexpected skip occurred when I clicked "Copy to Right/Left and Advance". I'll take a look at it later.

**Comment by sdottaka** [comment]
@pridmorej

I don't know if the problem you experienced has been resolved, but I tried to fix it with commit e7b2cae.
